### PR TITLE
Add support for MLIR printing/parsing of custom attributes 

### PR DESCRIPTION
**Pull request by math-fehr**
This PR lets us parse custom attributes in an MLIR files.
I added the custom parsing of LLVM struct, which should be the only custom attribute we have (the other being builtins).

### CODE REVIEW

File: src/xdsl/parser.py [hunk 1]
```diff
@@ -608,27 +605,48 @@ def parse_optional_attribute(self,
...
+
+        if int_type := self.try_parse(parse_integer_type):
+            return int_type
+
+        return None
+
+    def parse_optional_attribute(self,
+                                 skip_white_space: bool = True
+                                 ) -> Attribute | None:
+        if self.source == self.Source.MLIR:
```
Comment (webmiche) [inline src/xdsl/parser.py:622]: Maybe some comments here would help for the future. This should be MLIR costum parsing, the next is xDSL costum parsing, right?

File: tests/filecheck/mlir-conversion/llvm_test.xdsl [hunk 2]
```diff
@@ -1,29 +1,30 @@
...
   // CHECK-NEXT: }
 
-  func.func() ["sym_name" = "struct_to_struct2", "function_type" = !fun<[!llvm.struct<"", [!i32, !i32]>], [!llvm.struct<"", [!i32, !i32]>]>, "sym_visibility" = "private"] {
-    // CHECK: func private @struct_to_struct2(%arg0: !llvm.struct<(i32, i32)>) -> !llvm.struct<(i32, i32)> {
-    ^1(%1 : !llvm.struct<"", [!i32, !i32]>):
-      func.return(%1 : !llvm.struct<"", [!i32, !i32]>)
-    // CHECK-NEXT: return %arg0 : !llvm.struct<(i32, i32)>
+  func.func() ["sym_name" = "struct_to_struct2", "function_type" = !fun<[!llvm.struct<(!i32, !i32)>], [!llvm.struct<(!i32, !i32)>]>, "sym_visibility" = "private"] {
+    // CHECK: func private @struct_to_struct2(%{{.*}}: !llvm.struct<(i32, i32)>) -> !llvm.struct<(i32, i32)> {
+    ^1(%1 : !llvm.struct<(!i32, !i32)>):
```
Comment (webmiche) [inline tests/filecheck/mlir-conversion/llvm_test.xdsl:17]:
```suggestion
    ^1(%{{.*}} : !llvm.struct<(!i32, !i32)>):
```

Dito for later ones.

File: tests/filecheck/parser-printer/llvm_mlir_printer.xdsl [hunk 3]
```diff
@@ -1,17 +0,0 @@
-// RUN: xdsl-opt -t mlir %s | filecheck %s
```
Comment (webmiche) [inline tests/filecheck/parser-printer/llvm_mlir_printer.xdsl:1]: Why was this file removed? Do we not support this functionality anymore?

File: src/xdsl/dialects/llvm.py [hunk 4]
```diff
@@ -51,6 +51,13 @@ def print_parameters(self, printer: Printer) -> None:
         printer.print_list(self.types.data, printer.print_attribute)
         printer.print(")>")
 
+    @staticmethod
+    def parse_parameters(parser: Parser) -> list[Attribute]:
```
Comment (webmiche) [inline src/xdsl/dialects/llvm.py:55]: Notice that this parsing will have to change at some point, as the LLVMStructType can come in 3 different forms. However, that does not have to be part of this PR.

**Review by webmiche — APPROVED** [review]
LGTM! I feel like this PR could also use some testing with the python bindings, so something where you print with the python bindings and parse with our MLIR parser and the other way around.

File: tests/filecheck/mlir-conversion/llvm_test.xdsl (hunk 2 above)
Comment (math-fehr) [reply tests/filecheck/mlir-conversion/llvm_test.xdsl:17]: Ahah, I made this exact mistake when changing the test case, but here this SSA value is the value we are parsing, and not the value in a CHECK ;)

File: tests/filecheck/parser-printer/llvm_mlir_printer.xdsl (hunk 3 above)
Comment (math-fehr) [reply tests/filecheck/parser-printer/llvm_mlir_printer.xdsl:1]: This file was a subset of the `llvm_test.xdsl` one. So now this line is in `llvm_test.xdsl`

File: tests/filecheck/mlir-conversion/llvm_test.xdsl (hunk 2 above)
Comment (webmiche) [reply tests/filecheck/mlir-conversion/llvm_test.xdsl:17]: Ah yes true. I guess I haven't read enough file checks with inlined checking :sweat_smile:

**Review by webmiche — APPROVED** [review]
